pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,187 @@
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api"
+	coreinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions/core/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+type PersistentVolumeLabelController struct {
+	pvlInformer      coreinformers.PersistentVolumeInformer
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
+	cloud            cloudprovider.Interface
+	mutex            sync.Mutex
+}
+
+const ()
+
+// NewPersistentVolumeLabelController creates a PersistentVolumeLabelController object
+func NewPersistentVolumeLabelController(
+	pvlInformer coreinformers.PersistentVolumeInformer,
+	cloud cloudprovider.Interface) *PersistentVolumeLabelController {
+	pvlc := &PersistentVolumeLabelController{
+		pvlInformer: pvlInformer,
+		cloud:       cloud,
+	}
+
+	pvlInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: pvlc.AddLabels,
+	})
+	return pvlc
+}
+
+// This controller adds labels to persistent volumes and sets them to
+// an available state
+func (pvl *PersistentVolumeLabelController) Run(stopCh <-chan struct{}) {
+	go func() {
+		defer utilruntime.HandleCrash()
+		<-stopCh
+	}()
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) AddLabels(obj interface{}) {
+	volume, ok := obj.(*api.PersistentVolume)
+	if !ok {
+		return
+	}
+	var volumeLabels map[string]string
+	if volume.Spec.AWSElasticBlockStore != nil {
+		labels, err := pvlc.findAWSEBSLabels(volume)
+		if err != nil {
+			glog.Errorf("error querying AWS EBS volume %s: %v", volume.Spec.AWSElasticBlockStore.VolumeID, err)
+		} else {
+			volumeLabels = labels
+		}
+	}
+	if volume.Spec.GCEPersistentDisk != nil {
+		labels, err := pvlc.findGCEPDLabels(volume)
+		if err != nil {
+			glog.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+		} else {
+			volumeLabels = labels
+		}
+	}
+
+	if len(volumeLabels) != 0 {
+		if volume.Labels == nil {
+			volume.Labels = make(map[string]string)
+		}
+		for k, v := range volumeLabels {
+			// We (silently) replace labels if they are provided.
+			// This should be OK because they are in the kubernetes.io namespace
+			// i.e. we own them
+			volume.Labels[k] = v
+		}
+		volume.Status.Phase = api.VolumeAvailable
+	}
+}
+
+func (pvlc *PersistentVolumeLabelController) findAWSEBSLabels(volume *api.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.AWSElasticBlockStore.VolumeID == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+	ebsVolumes, err := pvlc.getEBSVolumes()
+	if err != nil {
+		return nil, err
+	}
+	if ebsVolumes == nil {
```

Comment:
Yes, but it is calling out to the cloud.  There are contingents in getEBSVolumes that would result in a nil being returned, namely on a cast issue.  Unlikely to occur, but original code accounted for it.